pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,253 @@
+	cmd.Flags().StringVar(&opts.ExpectedIdentity, "expected-identity", opts.ExpectedIdentity, "An expected image docker reference to verify.")
+	return cmd
+}
+
+func (o *VerifyImageSignatureOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+	if len(args) != 1 {
+		return kcmdutil.UsageError(cmd, "exactly one image must be specified")
+	}
+	o.InputImage = args[0]
+	if len(o.ExpectedIdentity) == 0 {
+		return kcmdutil.UsageError(cmd, "the --expected-identity must be specified")
+	}
+	var err error
+
+	// If --public-key is provided only this key will be used for verification and the
+	// .gnupg/pubring.gpg will be ignored.
+	if len(o.PublicKeyFilename) > 0 {
+		if o.Remove {
+			return kcmdutil.UsageError(cmd, "cannot use public key when removing verification status")
+		}
+		if o.PublicKey, err = ioutil.ReadFile(o.PublicKeyFilename); err != nil {
+			return err
+		}
+	}
+	if o.Client, _, err = f.Clients(); err != nil {
+		return err
+	}
+	// Only make this API call when we are sure we will be writing validation.
+	if o.Confirm && !o.Remove {
+		if me, err := o.Client.Users().Get("~"); err != nil {
+			return err
+		} else {
+			o.CurrentUser = me.Name
+		}
+	}
+
+	return nil
+}
+
+// verifySignature verifies the image signature and returns the identity when the signature
+// is valid.
+// TODO: This should be calling the 'containers/image' library in future.
+func (o *VerifyImageSignatureOptions) verifySignature(signature []byte) (string, []byte, error) {
+	var (
+		mechanism SigningMechanism
+		err       error
+	)
+	// If public key is specified, use JUST that key for verification. Otherwise use all
+	// keys in local GPG public keyring.
+	if len(o.PublicKeyFilename) == 0 {
+		mechanism, err = newGPGSigningMechanismInDirectory("")
+	} else {
+		mechanism, _, err = newEphemeralGPGSigningMechanism(o.PublicKey)
+	}
+	if err != nil {
+		return "", nil, err
+	}
+	defer mechanism.Close()
+	content, identity, err := mechanism.Verify(signature)
+	if err != nil {
+		return "", nil, err
+	}
+	return string(identity), content, nil
+}
+
+// verifySignatureContent verifies that the signature content matches the given image.
+// TODO: This should be done by calling the 'containers/image' library in future.
+func (o *VerifyImageSignatureOptions) verifySignatureContent(content []byte) (string, error) {
+	// TODO: The types here are just to decompose the JSON. The fields should not change but
+	// we need to use containers/image library here to guarantee compatibility in future.
+	type criticalImage struct {
+		Digest string `json:"docker-manifest-digest"`
+	}
+	type criticalIdentity struct {
+		DockerReference string `json:"docker-reference"`
+	}
+	type critical struct {
+		Image    criticalImage    `json:"image"`
+		Identity criticalIdentity `json:"identity"`
+	}
+	type message struct {
+		Critical critical `json:"critical"`
+	}
+	m := message{}
+	if err := json.Unmarshal(content, &m); err != nil {
```

Comment:
According to @mtrmac, you should first unmarshal into an `interface{}` and ensure it doesn't contain any unrecognized fields.

> Any unknown JSON field in critical, and any subobjects, must cause the signature to be rejected. That's what critical means in this format.